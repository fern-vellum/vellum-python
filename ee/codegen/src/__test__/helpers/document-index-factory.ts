export const mockDocumentIndexFactory = () => ({
  id: "b6ef8809-346e-469c-beed-2e5c4e9844c5",
  name: "my-sweet-document",
});
