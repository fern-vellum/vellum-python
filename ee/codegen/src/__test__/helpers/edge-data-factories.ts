import { WorkflowEdge, WorkflowNode } from "src/types/vellum";

const getSourceHandleId = (
  node: WorkflowNode | [WorkflowNode, string]
): string => {
  if (Array.isArray(node)) {
    const [actualNode, portName] = node;

    if (actualNode.type === "GENERIC") {
      const id = actualNode.ports.find((port) => port.name === portName)?.id;
      if (!id) {
        throw new Error(`Port not found: ${portName}`);
      }

      return id;
    }

    if (actualNode.type === "CONDITIONAL") {
      const id = actualNode.data.conditions[Number(portName)]?.sourceHandleId;
      if (!id) {
        throw new Error("Conditional node has no source handle id");
      }

      return id;
    }

    return getSourceHandleId(actualNode);
  }

  if (node.type === "GENERIC") {
    const port = node.ports[0];
    if (!port) {
      throw new Error("Generic node has no ports");
    }

    return port.id;
  }

  if (
    node.type == "TERMINAL" ||
    node.type === "NOTE" ||
    node.type === "ERROR"
  ) {
    throw new Error(`${node.type} nodes have no source handle id`);
  }

  if (node.type === "CONDITIONAL") {
    throw new Error("Conditional nodes must include a port name");
  }

  return node.data.sourceHandleId;
};

const getTargetHandleId = (
  node: WorkflowNode | [WorkflowNode, number]
): string => {
  if (Array.isArray(node)) {
    const [actualNode, targetIndex] = node;
    if (actualNode.type === "MERGE") {
      const id = actualNode.data.targetHandles[targetIndex]?.id;
      if (!id) {
        throw new Error(`Target handle not found: ${targetIndex}`);
      }

      return id;
    }

    return getTargetHandleId(actualNode);
  }

  if (node.type === "GENERIC") {
    return node.trigger.id;
  }

  if (node.type === "MERGE") {
    throw new Error("Merge nodes must include target handle index");
  }

  if (node.type === "ENTRYPOINT" || node.type === "NOTE") {
    throw new Error(`${node.type} nodes have no target handle id`);
  }

  return node.data.targetHandleId;
};

export type EdgeFactoryNodePair = [
  WorkflowNode | [WorkflowNode, string],
  WorkflowNode | [WorkflowNode, number]
];

export function edgesFactory(nodePairs: EdgeFactoryNodePair[]): WorkflowEdge[] {
  return nodePairs.map(([sourceNode, targetNode], index) => ({
    id: `edge-${index + 1}`,
    type: "DEFAULT",
    sourceNodeId: Array.isArray(sourceNode) ? sourceNode[0].id : sourceNode.id,
    sourceHandleId: getSourceHandleId(sourceNode),
    targetNodeId: Array.isArray(targetNode) ? targetNode[0].id : targetNode.id,
    targetHandleId: getTargetHandleId(targetNode),
  }));
}
