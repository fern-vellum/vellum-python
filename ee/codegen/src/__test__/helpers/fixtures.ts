import { readdirSync } from "fs";
import { basename, join, relative } from "path";

const fixtureDir = join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "tests",
  "codegen_integration",
  "fixtures"
);

const allFixtures = readdirSync(fixtureDir);

function getFixtures(
  excludeFixtures: Set<string> | string[] = new Set(),
  includeFixtures: Set<string> | string[] = new Set()
): string[] {
  const excludeSet = Array.isArray(excludeFixtures)
    ? new Set(excludeFixtures)
    : excludeFixtures;
  const includeSet = Array.isArray(includeFixtures)
    ? new Set(includeFixtures)
    : includeFixtures;

  return allFixtures.filter((f) => {
    const fileName = basename(f);

    return (
      (!excludeSet.size || !excludeSet.has(fileName)) &&
      (!includeSet.size || includeSet.has(fileName))
    );
  });
}

function getFixturePaths(root: string) {
  const displayFile = join(
    fixtureDir,
    root,
    "display_data",
    `${basename(root)}.json`
  );
  const codeDir = join(fixtureDir, root, "code");

  return { displayFile, codeDir };
}

export function getFixturesForProjectTest(
  {
    excludeFixtures,
    includeFixtures,
  }: {
    excludeFixtures?: Set<string> | string[];
    includeFixtures?: Set<string> | string[];
  } = { excludeFixtures: new Set(), includeFixtures: new Set() }
): {
  fixtureName: string;
  displayFile: string;
  codeDir: string;
}[] {
  return getFixtures(excludeFixtures, includeFixtures).map((root) => ({
    fixtureName: root,
    ...getFixturePaths(root),
  }));
}

export function getAllFilesInDir(
  dirPath: string,
  ignoreRegexes?: RegExp[]
): { [relativePath: string]: string } {
  const files: { [relativePath: string]: string } = {};

  function traverseDir(currentPath: string) {
    const entries = readdirSync(currentPath, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name);

      if (entry.isDirectory()) {
        traverseDir(fullPath);
      } else if (entry.isFile()) {
        const relativePath = relative(dirPath, fullPath);

        if (
          ignoreRegexes &&
          ignoreRegexes.some((regex) => regex.test(relativePath))
        ) {
          continue;
        }

        files[relativePath] = fullPath;
      }
    }
  }

  traverseDir(dirPath);
  return files;
}
