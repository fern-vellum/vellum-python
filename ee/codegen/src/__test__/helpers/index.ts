export { nodeContextFactory } from "./node-context-factory";
export { workflowContextFactory } from "./workflow-context-factory";
