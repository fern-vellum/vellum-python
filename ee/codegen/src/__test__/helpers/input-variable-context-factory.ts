import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { InputVariableContext } from "src/context/input-variable-context";

export function inputVariableContextFactory({
  inputVariableData,
  workflowContext,
}: {
  inputVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): InputVariableContext {
  return new InputVariableContext({
    inputVariableData,
    workflowContext,
  });
}
