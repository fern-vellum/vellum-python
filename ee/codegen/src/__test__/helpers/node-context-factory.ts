import { workflowContextFactory } from "./workflow-context-factory";

import { createNodeContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  SearchNode,
  WorkflowDataNode,
  WorkflowNodeType as WorkflowNodeTypeEnum,
} from "src/types/vellum";

export async function nodeContextFactory({
  workflowContext,
  nodeData,
}: Partial<BaseNodeContext.Args<WorkflowDataNode>> = {}): Promise<
  BaseNodeContext<WorkflowDataNode>
> {
  return await createNodeContext({
    workflowContext: workflowContext ?? workflowContextFactory(),
    nodeData: nodeData ?? {
      id: "search",
      type: WorkflowNodeTypeEnum.SEARCH,
      inputs: [],
      data: { label: "MyNode" } as SearchNode["data"],
    },
  });
}
