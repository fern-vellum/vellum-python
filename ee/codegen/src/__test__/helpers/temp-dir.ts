import os from "os";
import path from "path";

export function makeTempDir(base?: string): string {
  const randomSuffix = Math.random().toString(36).substring(7);
  return path.join(os.tmpdir(), `${base ?? "test-directory"}-${randomSuffix}`);
}
