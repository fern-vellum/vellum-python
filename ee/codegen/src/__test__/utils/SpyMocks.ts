import { MetricDefinitionHistoryItem } from "vellum-ai/api";
import { MetricDefinitions as MetricDefinitionsClient } from "vellum-ai/api/resources/metricDefinitions/client/Client";
import { MockInstance } from "vitest";

export class SpyMocks {
  static createMetricDefinitionMock(): MockInstance {
    return vi
      .spyOn(
        MetricDefinitionsClient.prototype,
        "metricDefinitionHistoryItemRetrieve"
      )
      .mockResolvedValue({
        id: "mocked-metric-output-id",
        label: "mocked-metric-output-label",
        name: "mocked-metric-output-name",
        description: "mocked-metric-output-description",
        outputVariables: [
          {
            id: "0e455862-ccc4-47a4-a9a5-061fadc94fd6",
            key: "score",
            type: "NUMBER",
          },
        ],
      } as MetricDefinitionHistoryItem);
  }
}
