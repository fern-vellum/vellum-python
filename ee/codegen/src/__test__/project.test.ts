import { mkdir, rm } from "fs/promises";
import * as fs from "node:fs";
import { join } from "path";

import { difference } from "lodash";
import { expect } from "vitest";

import {
  getAllFilesInDir,
  getFixturesForProjectTest,
} from "./helpers/fixtures";
import { makeTempDir } from "./helpers/temp-dir";

import { SpyMocks } from "src/__test__/utils/SpyMocks";
import { NodeAttributeGenerationError } from "src/generators/errors";
import { WorkflowProjectGenerator } from "src/project";

describe("WorkflowProjectGenerator", () => {
  let tempDir: string;

  beforeEach(async () => {
    tempDir = makeTempDir("project-test");
    await mkdir(tempDir, { recursive: true });
  });

  afterEach(async () => {
    await rm(tempDir, { recursive: true, force: true });
  });

  describe("generateCode", () => {
    const excludeFilesAtPaths: RegExp[] = [/\.pyc$/];
    const ignoreContentsOfFilesAtPaths: RegExp[] = [];
    const fixtureMocks = {
      simple_guard_rail_node: SpyMocks.createMetricDefinitionMock(),
      faa_q_and_a_bot: SpyMocks.createWorkflowDeploymentsMock(),
    };

    it.each(
      getFixturesForProjectTest({
        includeFixtures: [
          "simple_search_node",
          "simple_inline_subworkflow_node",
          "simple_guardrail_node",
          "simple_prompt_node",
          "simple_map_node",
          "simple_code_execution_node",
          "simple_conditional_node",
          "simple_templating_node",
          "simple_error_node",
          "simple_merge_node",
          "simple_api_node",
          "simple_node_with_try_wrap",
        ],
        fixtureMocks: fixtureMocks,
      })
    )(
      "should correctly generate code for fixture $fixtureName",
      async ({ displayFile, codeDir }) => {
        const displayData: unknown = JSON.parse(
          fs.readFileSync(displayFile, "utf-8")
        );

        const project = new WorkflowProjectGenerator({
          absolutePathToOutputDirectory: tempDir,
          workflowVersionExecConfigData: displayData,
          moduleName: "code",
          vellumApiKey: "<TEST_API_KEY>",
          options: {
            codeExecutionNodeCodeRepresentationOverride: "STANDALONE",
          },
        });

        await project.generateCode();

        const generatedFiles = getAllFilesInDir(
          join(tempDir, project.getModuleName())
        );
        const expectedFiles = getAllFilesInDir(codeDir, excludeFilesAtPaths);

        const extraFilePaths = difference(
          Object.keys(generatedFiles),
          Object.keys(expectedFiles)
        );
        const extraFiles = extraFilePaths.map((path) => generatedFiles[path]);
        expect(extraFiles.length, `Found extra file(s): ${extraFiles}`).toBe(0);

        for (const [
          expectedRelativePath,
          expectedAbsolutePath,
        ] of Object.entries(expectedFiles)) {
          const generatedAbsolutePath = generatedFiles[expectedRelativePath];

          if (!generatedAbsolutePath) {
            throw new Error(
              `Expected to have generated a file at the path: ${expectedRelativePath}`
            );
          }

          if (
            ignoreContentsOfFilesAtPaths.some((regex) =>
              regex.test(expectedRelativePath)
            )
          ) {
            continue;
          }

          const generatedFileContents = fs.readFileSync(
            generatedAbsolutePath,
            "utf-8"
          );

          expect(generatedFileContents).toMatchFileSnapshot(
            expectedAbsolutePath,
            `File contents don't match snapshot: ${expectedRelativePath}`
          );
        }
      }
    );
  });
  describe("failure cases", () => {
    const displayData = {
      workflow_raw_data: {
        nodes: [
          {
            id: "entry",
            type: "ENTRYPOINT",
            data: {
              label: "Entrypoint",
              source_handle_id: "entry_source",
              target_handle_id: "entry_target",
            },
            inputs: [],
          },
          {
            id: "bad_node",
            type: "TEMPLATING",
            data: {
              label: "Bad Node",
              template_node_input_id: "template",
              output_id: "output",
              output_type: "STRING",
              source_handle_id: "template_source",
              target_handle_id: "template_target",
            },
            inputs: [
              {
                id: "template",
                key: "template",
                value: {
                  combinator: "OR",
                  rules: [
                    {
                      type: "CONSTANT_VALUE",
                      data: {
                        type: "STRING",
                        value: "foo",
                      },
                    },
                  ],
                },
              },
              {
                id: "input",
                key: "other",
                value: {
                  combinator: "OR",
                  rules: [
                    {
                      type: "NODE_OUTPUT",
                      data: {
                        node_id: "node_that_doesnt_exist",
                        output_id: "output",
                      },
                    },
                  ],
                },
              },
            ],
          },
        ],
        edges: [
          {
            source_node_id: "entry",
            source_handle_id: "entry_source",
            target_node_id: "bad_node",
            target_handle_id: "template_target",
            type: "DEFAULT",
            id: "edge_1",
          },
        ],
      },
      input_variables: [],
      output_variables: [],
    };

    it("should generate code even if a node fails to generate", async () => {
      const project = new WorkflowProjectGenerator({
        absolutePathToOutputDirectory: tempDir,
        workflowVersionExecConfigData: displayData,
        moduleName: "code",
        vellumApiKey: "<TEST_API_KEY>",
      });

      await project.generateCode();

      expect(
        fs.existsSync(join(tempDir, project.getModuleName(), "workflow.py"))
      ).toBe(true);
      expect(
        fs.existsSync(join(tempDir, project.getModuleName(), "nodes"))
      ).toBe(true);

      const badNodePath = join(
        tempDir,
        project.getModuleName(),
        "nodes",
        "bad_node.py"
      );
      expect(fs.existsSync(badNodePath)).toBe(true);
      expect(fs.readFileSync(badNodePath, "utf-8")).toBe(`\
from vellum.workflows.nodes.displayable import TemplatingNode
from vellum.workflows.state import BaseState


class BadNode(TemplatingNode[BaseState, str]):
    template = """foo"""
    inputs = {}
`);

      const errorLogPath = join(tempDir, project.getModuleName(), "error.log");
      expect(fs.existsSync(errorLogPath)).toBe(true);
      expect(fs.readFileSync(errorLogPath, "utf-8")).toBe(`\
Encountered 1 error(s) while generating code:

- Failed to generate attribute 'BadNode.inputs.other': Failed to find node with id 'node_that_doesnt_exist'
`);
    });

    it("should fail to generate code if a node fails in strict mode", async () => {
      const project = new WorkflowProjectGenerator({
        absolutePathToOutputDirectory: tempDir,
        workflowVersionExecConfigData: displayData,
        moduleName: "code",
        vellumApiKey: "<TEST_API_KEY>",
        strict: true,
      });

      expect(project.generateCode()).rejects.toThrow(
        new NodeAttributeGenerationError(
          "Failed to generate attribute 'BadNode.inputs.other': Failed to find node with id 'node_that_doesnt_exist'"
        )
      );
    });
  });
  describe("inlude sandbox", () => {
    const displayData = {
      workflow_raw_data: {
        nodes: [
          {
            id: "entry",
            type: "ENTRYPOINT",
            data: {
              label: "Entrypoint",
              source_handle_id: "entry_source",
              target_handle_id: "entry_target",
            },
            inputs: [],
          },
          {
            id: "templating-node",
            type: "TEMPLATING",
            data: {
              label: "Templating Node",
              template_node_input_id: "template",
              output_id: "output",
              output_type: "STRING",
              source_handle_id: "template_source",
              target_handle_id: "template_target",
            },
            inputs: [
              {
                id: "template",
                key: "template",
                value: {
                  combinator: "OR",
                  rules: [
                    {
                      type: "CONSTANT_VALUE",
                      data: {
                        type: "STRING",
                        value: "{{input}}",
                      },
                    },
                  ],
                },
              },
              {
                id: "input",
                key: "other",
                value: {
                  combinator: "OR",
                  rules: [
                    {
                      type: "INPUT_VARIABLE",
                      data: {
                        input_variable_id: "workflow-input",
                      },
                    },
                  ],
                },
              },
            ],
          },
        ],
        edges: [
          {
            source_node_id: "entry",
            source_handle_id: "entry_source",
            target_node_id: "templating-node",
            target_handle_id: "template_target",
            type: "DEFAULT",
            id: "edge_1",
          },
        ],
      },
      input_variables: [
        {
          key: "input",
          type: "STRING",
          id: "workflow-input",
        },
        {
          key: "chat",
          type: "CHAT_HISTORY",
          id: "workflow-chat",
        },
      ],
      output_variables: [],
    };
    it("should include a sandbox.py file when passed sandboxInputs", async () => {
      const project = new WorkflowProjectGenerator({
        absolutePathToOutputDirectory: tempDir,
        workflowVersionExecConfigData: displayData,
        moduleName: "code",
        vellumApiKey: "<TEST_API_KEY>",
        sandboxInputs: [
          [
            {
              name: "input",
              type: "STRING",
              value: "foo",
            },
            {
              name: "chat",
              type: "CHAT_HISTORY",
              value: [
                {
                  role: "USER",
                  text: "foo",
                },
              ],
            },
          ],
          [
            {
              name: "input",
              type: "STRING",
              value: "bar",
            },
            {
              name: "chat",
              type: "CHAT_HISTORY",
              value: [
                {
                  role: "USER",
                  content: {
                    type: "STRING",
                    value: "bar",
                  },
                  // The API sometimes sends null for source, but it's not a valid value
                  source: null as unknown as undefined,
                },
              ],
            },
          ],
          [
            {
              name: "input",
              type: "STRING",
              value: "hello",
            },
            {
              name: "chat",
              type: "CHAT_HISTORY",
              value: [
                {
                  role: "USER",
                  content: {
                    type: "ARRAY",
                    value: [
                      {
                        type: "STRING",
                        value: "hello",
                      },
                    ],
                  },
                },
              ],
            },
          ],
        ],
      });

      await project.generateCode();

      const sandboxPath = join(tempDir, project.getModuleName(), "sandbox.py");
      expect(fs.existsSync(sandboxPath)).toBe(true);
      expect(fs.readFileSync(sandboxPath, "utf-8")).toMatchSnapshot();
    });
  });
});
