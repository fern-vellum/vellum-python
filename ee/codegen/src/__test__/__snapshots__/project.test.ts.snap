// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkflowProjectGenerator > inlude sandbox > should include a sandbox.py file when passed sandboxInputs 1`] = `
"from vellum import ArrayChatMessageContent, ChatMessage, StringChatMessageContent
from vellum.workflows.sandbox import SandboxRunner

from .inputs import Inputs
from .workflow import Workflow

if __name__ != "__main__":
    raise Exception("This file is not meant to be imported")


runner = SandboxRunner(
    workflow=Workflow,
    inputs=[
        Inputs(
            input="foo",
            chat=[
                ChatMessage(role="USER", text="foo"),
            ],
        ),
        Inputs(
            input="bar",
            chat=[
                ChatMessage(role="USER", content=StringChatMessageContent(value="bar")),
            ],
        ),
        Inputs(
            input="hello",
            chat=[
                ChatMessage(
                    role="USER",
                    content=ArrayChatMessageContent(
                        value=[
                            StringChatMessageContent(value="hello"),
                        ]
                    ),
                ),
            ],
        ),
    ],
)

runner.run()
"
`;
